aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -145,23 +145,25 @@ void SparseCsrTensorImpl::resize_and_clear_(int64_t sparse_dim, IntArrayRef size
   refresh_numel();
 }
 
-void SparseCsrTensorImpl::resize_as_sparse_csr_tensor_(const Tensor& src) {
+void SparseCsrTensorImpl::resize_as_sparse_compressed_tensor_(
+    const Tensor& src) {
   TORCH_CHECK(
       !has_symbolic_sizes_strides_,
-      "resize_as_sparse_csr_tensor_ called on tensor with symbolic shape");
+      "resize_as_sparse_compressed_tensor_ called on tensor with symbolic shape");
   set_layout(src.layout());
-  crow_indices_ = at::empty_like(
-      src.crow_indices(),
-      src.crow_indices().options(),
-      src.crow_indices().suggest_memory_format());
-  col_indices_ = at::empty_like(
-      src.col_indices(),
-      src.col_indices().options(),
-      src.col_indices().suggest_memory_format());
-  values_ = at::empty_like(
-      src.values(),
-      src.values().options(),
-      src.values().suggest_memory_format());
+  Tensor compressed_indices;
+  Tensor plain_indices;
+  std::tie(compressed_indices, plain_indices) =
+      sparse_csr::getCompressedPlainIndices(src);
+
+  // NOTE: clone indices as empty_like will produce a tensor which is not valid
+  // under the invariants compressed indices are always crow_indices_ on the
+  // impl
+  crow_indices_ = compressed_indices.to(crow_indices_.options());
+  // plain indices are always col_indices_ on the impl
+  col_indices_ = plain_indices.to(col_indices_.options());
+  // Values are initializd empty
+  values_ = at::empty(src.values().sizes(), values_.options());
```

Comment:
The problem here is that `to` might be a no-op and return the same indices as `src`. Is that safe? Maybe we it is safer to do `clone().to(...)`. Also, `src.values()` will fail when `src` is not coalesced:
```python
In [1]: import torch

In [2]: idx = torch.zeros(2, 3, dtype=torch.long)

In [3]: val = torch.rand(3, 3, 3)

In [4]: s = torch.sparse_coo_tensor(idx, val, size=(10, 10, 3, 3))

In [5]: s.indices()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 s.indices()

RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first

In [6]: s.values()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 s.values()

RuntimeError: Cannot get values on an uncoalesced tensor, please call .coalesce() first

```

Reply:
@nikitaved , your examples use COO tensor, here we discuss CSR/... tensors.

Reply:
See https://github.com/pytorch/pytorch/pull/85378#discussion_r977498284 that ought to resolve various issues here.